File: packages/runtime-python/src/eventgen_runtime/adapters/message_memory.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from eventgen_runtime.types import EventEnvelope, ReceiveOptions


@dataclass
class InMemoryReceivedRecord:
    envelope: EventEnvelope
    attempts: int = 0
    acked: bool = False
    dead_lettered: bool = False


class _InMemoryReceivedMessage:
    def __init__(self, record: InMemoryReceivedRecord, source: "InMemoryMessageSource") -> None:
        self.envelope = record.envelope
        self.raw = record
        self.attributes: dict[str, Any] | None = None
        self._record = record
        self._source = source

    async def ack(self) -> None:
        self._record.acked = True

    async def retry(self, error: Any | None = None) -> None:
        return None

    async def dead_letter(self, error: Any | None = None) -> None:
        self._record.dead_lettered = True
        self._source.dead_letters.append(self._record)


class InMemoryMessageSource:
    def __init__(self, envelopes: list[EventEnvelope] | None = None) -> None:
        self.messages = [InMemoryReceivedRecord(envelope=envelope) for envelope in envelopes or []]
        self.dead_letters: list[InMemoryReceivedRecord] = []

    def enqueue(self, envelope: EventEnvelope) -> None:
        self.messages.append(InMemoryReceivedRecord(envelope=envelope))

    async def receive(self, options: ReceiveOptions | None = None) -> list[_InMemoryReceivedMessage]:
        max_messages = max(1, options.max_messages if options and options.max_messages else 1)
        records = [record for record in self.messages if not record.acked and not record.dead_lettered][:max_messages]
        for record in records:
            record.attempts += 1
        return [_InMemoryReceivedMessage(record, self) for record in records]

    async def close(self) -> None:
        return None

    def clear(self) -> None:
        self.messages.clear()
        self.dead_letters.clear()

File: packages/runtime-go/adapters/message_memory.go
package adapters

import (
	"context"

	runtime "github.com/eventgen/runtime-go"
)

type InMemoryReceivedRecord struct {
	Event        runtime.EventEnvelope
	Attempts     int
	Acked        bool
	DeadLettered bool
}

type InMemoryMessageSource struct {
	Messages    []*InMemoryReceivedRecord
	DeadLetters []*InMemoryReceivedRecord
}

func NewInMemoryMessageSource(envelopes ...runtime.EventEnvelope) *InMemoryMessageSource {
	source := &InMemoryMessageSource{}
	for _, envelope := range envelopes {
		source.Enqueue(envelope)
	}
	return source
}

func (s *InMemoryMessageSource) Enqueue(envelope runtime.EventEnvelope) {
	s.Messages = append(s.Messages, &InMemoryReceivedRecord{Event: envelope})
}

func (s *InMemoryMessageSource) Receive(_ context.Context, opts runtime.ReceiveOptions) ([]runtime.ReceivedMessage, error) {
	maxMessages := opts.MaxMessages
	if maxMessages <= 0 {
		maxMessages = 1
	}
	messages := make([]runtime.ReceivedMessage, 0, maxMessages)
	for _, record := range s.Messages {
		if len(messages) == maxMessages {
			break
		}
		if record.Acked || record.DeadLettered {
			continue
		}
		record.Attempts++
		messages = append(messages, &inMemoryReceivedMessage{source: s, record: record})
	}
	return messages, nil
}

func (s *InMemoryMessageSource) Close(_ context.Context) error { return nil }

type inMemoryReceivedMessage struct {
	source *InMemoryMessageSource
	record *InMemoryReceivedRecord
}

func (m *inMemoryReceivedMessage) Envelope() runtime.EventEnvelope { return m.record.Event }
func (m *inMemoryReceivedMessage) Raw() any                        { return m.record }
func (m *inMemoryReceivedMessage) Attributes() map[string]any      { return nil }
func (m *inMemoryReceivedMessage) Ack(_ context.Context) error {
	m.record.Acked = true
	return nil
}
func (m *inMemoryReceivedMessage) Retry(_ context.Context, _ error) error { return nil }
func (m *inMemoryReceivedMessage) DeadLetter(_ context.Context, _ error) error {
	m.record.DeadLettered = true
	m.source.DeadLetters = append(m.source.DeadLetters, m.record)
	return nil
}
